frontend/src/components/NewRunParameters.tsx

```diff
@@ -39,23 +61,145 @@ class NewRunParameters extends React.Component<NewRunParametersProps> {
     );
   }
 }
 
+interface EnhancedTextFieldProps {
+  param: ApiParameter;
+  index: number;
+  handleParamChange: (index: number, value: string) => void;
+}
+
+interface EnhancedTextFieldState {
+  isEditorOpen: boolean;
+  isInJsonForm: boolean;
+  isJsonField: boolean;
+}
+
+class EnhancedTextField extends React.Component<EnhancedTextFieldProps, EnhancedTextFieldState> {
+  public static getDerivedStateFromProps(
+    nextProps: EnhancedTextFieldProps,
+    prevState: EnhancedTextFieldState,
+  ): EnhancedTextFieldState {
+    let isJson = true;
+    try {
+      const displayValue = JSON.parse(nextProps.param.value || '');
+      // Nulls, booleans, strings, and numbers can all be parsed as JSON, but we don't care
+      // about rendering. Note that `typeOf null` returns 'object'
+      if (displayValue === null || typeof displayValue !== 'object') {
+        throw new Error('Parsed JSON was neither an array nor an object. Using default renderer');
+      }
+    } catch (err) {
+      isJson = false;
+    }
+    return {
+      isEditorOpen: prevState.isEditorOpen,
+      isInJsonForm: isJson,
+      isJsonField: prevState.isJsonField || isJson,
+    };
+  }
+
+  constructor(props: any) {
+    super(props);
+
+    this.state = {
+      isEditorOpen: false,
+      isInJsonForm: false,
+      isJsonField: false,
+    };
+  }
+
+  public render(): JSX.Element | null {
+    const { param, index, handleParamChange } = this.props;
+
+    if (this.state.isJsonField) {
+      return (
+        <>
+          <TextField
+            id={`newRunPipelineParam${index}`}
+            disabled={this.state.isEditorOpen}
+            variant='outlined'
+            label={param.name}
+            value={param.value || ''}
+            onChange={ev => handleParamChange(index, ev.target.value || '')}
+            style={{ maxWidth: 600 }}
+            className={commonCss.textField}
+            InputProps={{
+              classes: { disabled: css.nonEditableInput },
+              endAdornment: (
+                <InputAdornment position='end'>
+                  <Button
+                    color='secondary'
+                    id='chooseExperimentBtn'
+                    onClick={() => {
+                      if (this.state.isInJsonForm) {
+                        const displayValue = JSON.parse(param.value || '');
+                        if (this.state.isEditorOpen) {
+                          handleParamChange(index, JSON.stringify(displayValue) || '');
+                        } else {
+                          handleParamChange(index, JSON.stringify(displayValue, null, 2) || '');
+                        }
+                      }
+                      this.setState({
+                        isEditorOpen: !this.state.isEditorOpen,
+                      });
+                    }}
+                    style={{ padding: '3px 5px', margin: 0 }}
+                  >
+                    {this.state.isEditorOpen ? 'Close Editor' : 'Open Editor'}
```

Comment:
I'm not sure on this one. To be accurate, this is a JSON editor, not a general one. Do you think it's worth clarifying that?

Reply:
I think clarifying would help the user understand why the button is popping up, however it looks a little clunky. For this I think you are right and the value of an easy to understand design outweighs the cost of it looking somewhat clunky. I'll add it for now and can remove it if you change your mind

Reply:
Sounds good, I still prefer clarity:
1. it's a new feature
2. the button is hidden initially, only when parameter is JSON shows it. After the button is shown, opening the editor would be the best way to edit JSON. so it doesn't matter how much space the button takes.